# Предобработка данных


In [332]:
import numpy as np
import pandas as pd

In [333]:
DATASET_PATH = 'housing.csv'
PREPARED_DATASET_PATH = 'housing_prepared.csv'

## 1. Загрузка данных

#### **Описание задачи**
Цель - предсказать стоимость дома

#### **Описание датасета**

Статистические данные о ряде домов в Калифорнии, основанные на переписи 1990 года.

longitude - долгота

latitude - широта

housing_median_age - средний возраст дома

total_rooms - общее количество комнат

total_bedrooms - общее количество спален

population - количество проживающих

households - домохозяйства (семья)

ocean_proximity - близость океана

median_income - средний доход

median_house_value - средняя стоимость дома

In [334]:
? pd.read_csv

Signature:
 pd.read_csv(
    filepath_or_buffer: 'FilePath | ReadCsvBuffer[bytes] | ReadCsvBuffer[str]',
    *,
    sep: 'str | None | lib.NoDefault' = <no_default>,
    delimiter: 'str | None | lib.NoDefault' = None,
    header: "int | Sequence[int] | None | Literal['infer']" = 'infer',
    names: 'Sequence[Hashable] | None | lib.NoDefault' = <no_default>,
    index_col: 'IndexLabel | Literal[False] | None' = None,
    usecols: 'UsecolsArgType' = None,
    dtype: 'DtypeArg | None' = None,
    engine: 'CSVEngine | None' = None,
    converters: 'Mapping[Hashable, Callable] | None' = None,
    true_values: 'list | None' = None,
    false_values: 'list | None' = None,
    skipinitialspace: 'bool' = False,
    skiprows: 'list[int] | int | Callable[[Hashable], bool] | None' = None,
    skipfooter: 'int' = 0,
    nrows: 'int | None' = None,
    na_values: 'Hashable | Iterable[Hashable] | Mapping[Hashable, Iterable[Hashable]] | None' = None,
    keep_default_na: 'bool' = True,
    na_filter: 

In [335]:
df = pd.read_csv(DATASET_PATH)
df.head(10)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   
5    -122.25     37.85                52.0        919.0           213.0   
6    -122.25     37.84                52.0       2535.0           489.0   
7    -122.25     37.84                52.0       3104.0           687.0   
8    -122.26     37.84                42.0       2555.0           665.0   
9    -122.25     37.84                52.0       3549.0           707.0   

   population  households  median_income  median_house_value ocean_proximity  \
0       322.0       126.0         8.3252            452600.0        NEAR BAY   
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY   
2       496.0       177.0         7.2574            352100.0        NEAR BAY   
3       558.0       219.0         5.6431            341300.0        NEAR BAY   
4       565.0       259.0         3.8462            342200.0        NEAR BAY   
5       413.0       193.0         4.0368            269700.0        NEAR BAY   
6      1094.0       514.0         3.6591            299200.0        NEAR BAY   
7      1157.0       647.0         3.1200            241400.0        NEAR BAY   
8      1206.0       595.0         2.0804            226700.0        NEAR BAY   
9      1551.0       714.0         3.6912            261100.0        NEAR BAY   

   id  
0   0  
1   1  
2   2  
3   3  
4   4  
5   5  
6   6  
7   7  
8   8  
9   9

In [336]:
df.shape

(20640, 11)

In [337]:
df.columns

Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'median_house_value', 'ocean_proximity', 'id'],
      dtype='object')

In [338]:
df.index

RangeIndex(start=0, stop=20640, step=1)

In [339]:
df[['longitude', 'latitude']].head(2)

longitude  latitude
0    -122.23     37.88
1    -122.22     37.86

In [340]:
df[df['longitude'] > 10].head(2)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
3479     118.51     34.29                29.0       1287.0           194.0   
5904     118.43     34.29                39.0       1769.0           410.0   

      population  households  median_income  median_house_value  \
3479       525.0       187.0         6.4171            319300.0   
5904      1499.0       390.0         3.1212            153500.0   

     ocean_proximity    id  
3479       <1H OCEAN  3479  
5904       <1H OCEAN  5904

In [341]:
df[(df['longitude'] > 10) | (df['latitude'] < 100)].head(2)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   

   population  households  median_income  median_house_value ocean_proximity  \
0       322.0       126.0         8.3252            452600.0        NEAR BAY   
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY   

   id  
0   0  
1   1

In [342]:
df[(df['longitude'] > 10) & (df['latitude'] < 100)].head(2)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
3479     118.51     34.29                29.0       1287.0           194.0   
5904     118.43     34.29                39.0       1769.0           410.0   

      population  households  median_income  median_house_value  \
3479       525.0       187.0         6.4171            319300.0   
5904      1499.0       390.0         3.1212            153500.0   

     ocean_proximity    id  
3479       <1H OCEAN  3479  
5904       <1H OCEAN  5904

In [343]:
df.loc[df['longitude'] > 10, 'median_income'].head(2)

3479    6.4171
5904    3.1212
Name: median_income, dtype: float64

In [344]:
df.loc[df['longitude'] > 10, ['median_income', 'median_house_value']].head(2)

median_income  median_house_value
3479         6.4171            319300.0
5904         3.1212            153500.0

## 2. Приведение типов данных

In [345]:
df.dtypes

longitude             float64
latitude              float64
housing_median_age    float64
total_rooms           float64
total_bedrooms        float64
population            float64
households            float64
median_income         float64
median_house_value    float64
ocean_proximity        object
id                      int64
dtype: object

In [346]:
type(df['id'])

pandas.core.series.Series

In [347]:
type(df.id)

pandas.core.series.Series

In [348]:
type(df['id'].values)

numpy.ndarray

In [349]:
df['id'].dtype

dtype('int64')

In [350]:
df['id'] = df['id'].astype(str)
df['id'].dtype

dtype('O')

### Обзор количественных переменных

In [351]:
df_num_features = df.select_dtypes(include=['float64', 'int64'])
df_num_features.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value  
0       322.0       126.0         8.3252            452600.0  
1      2401.0      1138.0         8.3014            358500.0  
2       496.0       177.0         7.2574            352100.0  
3       558.0       219.0         5.6431            341300.0  
4       565.0       259.0         3.8462            342200.0

In [352]:
df.shape

(20640, 11)

In [353]:
df.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        19918.000000  20640.000000   
mean    -119.471242     35.036934           28.653630   2635.763081   
std        5.041408     94.903955           12.576796   2181.615252   
min     -124.350000 -13534.030000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max      122.030000   1327.130000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20041.000000  20640.000000   20640.000000   
mean       537.870553   1425.418243    499.539680       3.870671   
std        421.385070   1135.185798    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        296.000000    786.000000    280.000000       2.563400   
50%        435.000000   1165.000000    409.000000       3.534800   
75%        647.000000   1726.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        20640.000000  
mean        206855.816909  
std         115395.615874  
min          14999.000000  
25%         119600.000000  
50%         179700.000000  
75%         264725.000000  
max         500001.000000

In [354]:
df.describe().T

count           mean            std         min  \
longitude           20640.0    -119.471242       5.041408   -124.3500   
latitude            20640.0      35.036934      94.903955 -13534.0300   
housing_median_age  19918.0      28.653630      12.576796      1.0000   
total_rooms         20640.0    2635.763081    2181.615252      2.0000   
total_bedrooms      20433.0     537.870553     421.385070      1.0000   
population          20041.0    1425.418243    1135.185798      3.0000   
households          20640.0     499.539680     382.329753      1.0000   
median_income       20640.0       3.870671       1.899822      0.4999   
median_house_value  20640.0  206855.816909  115395.615874  14999.0000   

                            25%          50%           75%          max  
longitude             -121.8000    -118.4900    -118.01000     122.0300  
latitude                33.9300      34.2600      37.71000    1327.1300  
housing_median_age      18.0000      29.0000      37.00000      52.0000  
total_rooms           1447.7500    2127.0000    3148.00000   39320.0000  
total_bedrooms         296.0000     435.0000     647.00000    6445.0000  
population             786.0000    1165.0000    1726.00000   35682.0000  
households             280.0000     409.0000     605.00000    6082.0000  
median_income            2.5634       3.5348       4.74325      15.0001  
median_house_value  119600.0000  179700.0000  264725.00000  500001.0000

Пример расчета статистик

In [355]:
x = np.array([1, 3, 2, 4, 5, 6])
x = np.sort(x)
x

array([1, 2, 3, 4, 5, 6])

In [356]:
np.quantile(x, q=0.5)

np.float64(3.5)

In [357]:
np.quantile(df['total_rooms'], q=0.5)

np.float64(2127.0)

### Обзор категориальных переменных

In [358]:
df_obj_features = df.select_dtypes(include='object')
df_obj_features.head()

ocean_proximity id
0        NEAR BAY  0
1        NEAR BAY  1
2        NEAR BAY  2
3        NEAR BAY  3
4        NEAR BAY  4

In [359]:
df['ocean_proximity'].value_counts()

ocean_proximity
<1H OCEAN     9127
INLAND        6542
NEAR OCEAN    2655
NEAR BAY      2288
-               23
ISLAND           5
Name: count, dtype: int64

In [360]:
df['ocean_proximity'].unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', '-', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

In [361]:
df['ocean_proximity'].nunique()

6

## 3. Обработка пропусков

In [362]:
df.shape[0]

20640

In [363]:
df.isnull().sum()

longitude               0
latitude                0
housing_median_age    722
total_rooms             0
total_bedrooms        207
population            599
households              0
median_income           0
median_house_value      0
ocean_proximity         0
id                      0
dtype: int64

**housing_median_age**

In [364]:
median = df['housing_median_age'].median()
df['housing_median_age'] = df['housing_median_age'].fillna(median)
df.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population            599
households              0
median_income           0
median_house_value      0
ocean_proximity         0
id                      0
dtype: int64

In [365]:
median

np.float64(29.0)

**total_bedrooms**

In [366]:
median = df['total_bedrooms'].median()
df['total_bedrooms'] = df['total_bedrooms'].fillna(median)

In [367]:
df.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms          0
population            599
households              0
median_income           0
median_house_value      0
ocean_proximity         0
id                      0
dtype: int64

**population**

In [368]:
median = df['population'].median()
df['population'] = df['population'].fillna(median)
df.isnull().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
ocean_proximity       0
id                    0
dtype: int64

***Все и сразу***

In [369]:
median = df[['population', 'housing_median_age', 'total_bedrooms']].median()
median

population            1165.0
housing_median_age      29.0
total_bedrooms         435.0
dtype: float64

In [370]:
df[['population', 'housing_median_age', 'total_bedrooms']] = df[['population', 'housing_median_age', 'total_bedrooms']].fillna(median)

In [371]:
df.isnull().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
ocean_proximity       0
id                    0
dtype: int64

**ocean_proximity**


In [372]:
df['ocean_proximity'].mode()

0    <1H OCEAN
Name: ocean_proximity, dtype: object

In [373]:
df['ocean_proximity'].mode()[0]

'<1H OCEAN'

In [374]:
df.replace({'ocean_proximity': {'-': df['ocean_proximity'].mode()[0]}}, inplace=True)

In [375]:
df['ocean_proximity'].value_counts()

ocean_proximity
<1H OCEAN     9150
INLAND        6542
NEAR OCEAN    2655
NEAR BAY      2288
ISLAND           5
Name: count, dtype: int64

## 4. Обработка выбросов

In [376]:
df.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        20640.000000  20640.000000   
mean    -119.471242     35.036934           28.665746   2635.763081   
std        5.041408     94.903955           12.355019   2181.615252   
min     -124.350000 -13534.030000            1.000000      2.000000   
25%     -121.800000     33.930000           19.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max      122.030000   1327.130000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20640.000000  20640.000000  20640.000000   20640.000000   
mean       536.838857   1417.860562    499.539680       3.870671   
std        419.391878   1119.445348    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        297.000000    797.000000    280.000000       2.563400   
50%        435.000000   1165.000000    409.000000       3.534800   
75%        643.250000   1701.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        20640.000000  
mean        206855.816909  
std         115395.615874  
min          14999.000000  
25%         119600.000000  
50%         179700.000000  
75%         264725.000000  
max         500001.000000

**longitude**

-180 <= longitude <= 180

Положительные - восточная долгота, отрицательные - западная долгота

Калифорния - западная долгота

In [377]:
df[df['longitude'] >= 0]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
3479      118.51     34.29                29.0       1287.0           194.0   
5904      118.43     34.29                39.0       1769.0           410.0   
8405      118.36     33.93                40.0       1625.0           500.0   
8636      118.41     33.88                43.0       2492.0           449.0   
13051     121.29     38.61                17.0      13553.0          2474.0   
15263     117.27     33.02                21.0       2144.0           340.0   
17085       0.00     37.47                33.0       1266.0           415.0   
17359       0.00     34.88                 4.0       3680.0           559.0   
18551     122.03     36.96                28.0       1607.0           421.0   
19423       0.00     37.69                 5.0       9601.0          1639.0   

       population  households  median_income  median_house_value  \
3479        525.0       187.0         6.4171            319300.0   
5904       1499.0       390.0         3.1212            153500.0   
8405       2036.0       476.0         2.6298            156500.0   
8636       1033.0       437.0         7.9614            500001.0   
13051      6544.0      2359.0         3.9727            132700.0   
15263       928.0       344.0         5.7980            286100.0   
17085      1991.0       334.0         2.9200            202800.0   
17359      1678.0       569.0         5.0639            201700.0   
18551       926.0       385.0         2.4250            216100.0   
19423      4449.0      1575.0         4.5332            195500.0   

      ocean_proximity     id  
3479        <1H OCEAN   3479  
5904        <1H OCEAN   5904  
8405        <1H OCEAN   8405  
8636        <1H OCEAN   8636  
13051          INLAND  13051  
15263      NEAR OCEAN  15263  
17085      NEAR OCEAN  17085  
17359       <1H OCEAN  17359  
18551      NEAR OCEAN  18551  
19423          INLAND  19423

In [378]:
df.loc[df['longitude'] > 0, 'longitude'] * -1

3479    -118.51
5904    -118.43
8405    -118.36
8636    -118.41
13051   -121.29
15263   -117.27
18551   -122.03
Name: longitude, dtype: float64

In [379]:
df.loc[df['longitude'] > 0, 'longitude'] = df.loc[df['longitude'] > 0, 'longitude'] * -1

In [380]:
df.loc[df['longitude'] == 0, 'longitude'] = df['longitude'].median()

In [381]:
df[df['longitude'] >= 0]

Empty DataFrame
Columns: [longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, median_house_value, ocean_proximity, id]
Index: []

**latitude**

In [382]:
df[(df['latitude'] <= 0) | (df['latitude'] > 50)]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
8283     -118.13 -13534.03                45.0       1016.0           172.0   
12772    -121.42   1327.13                29.0       2217.0           536.0   

       population  households  median_income  median_house_value  \
8283        361.0       163.0         7.5000            434500.0   
12772      1203.0       507.0         1.9412             73100.0   

      ocean_proximity     id  
8283       NEAR OCEAN   8283  
12772          INLAND  12772

Калифорния вытянута вдоль берега Тихого Океана между 32 и 42 гр. северной широты 114 и 124 западной долготы.

In [383]:
df.loc[(df['latitude'] <= 0) | (df['latitude'] > 50), 'latitude'] = df['latitude'].median()

## 5. Отбор и построение новых признаков (фичей)

Исключаем признак "id"

In [384]:
df = df[df.columns[:-1]]
df

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1165.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  
0            NEAR BAY  
1            NEAR BAY  
2            NEAR BAY  
3            NEAR BAY  
4            NEAR BAY  
...               ...  
20635          INLAND  
20636          INLAND  
20637          INLAND  
20638          INLAND  
20639          INLAND  

[20640 rows x 10 columns]

### 5.1 Количественные переменные

In [385]:
# Доля спален в общем кол-ве комнат
df['bedroom_share'] = df['total_bedrooms'] / df['total_rooms'] * 100

# Сколько человек в среднем живут в одной комнате 
df['population_per_room'] = df['population'] / df['total_rooms']

In [386]:
df

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1165.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  bedroom_share  population_per_room  
0            NEAR BAY      14.659091             0.365909  
1            NEAR BAY      15.579659             0.338217  
2            NEAR BAY      12.951602             0.338105  
3            NEAR BAY      18.445840             0.437991  
4            NEAR BAY      17.209588             0.347265  
...               ...            ...                  ...  
20635          INLAND      22.462462             0.507508  
20636          INLAND      21.520803             0.510760  
20637          INLAND      21.517303             0.516859  
20638          INLAND      21.989247             0.398387  
20639          INLAND      22.118492             0.498025  

[20640 rows x 12 columns]

#### 5.2 Категориальные переменные

In [387]:
for i in df.columns: # перебираем все столбцы
    if str(df[i].dtype) == 'object': # если тип столбца - object
        print('=' * 10)
        print(i) # выводим название столбца
        print(set(df[i])) # выводим set значений этого столбца
        print('\n')

ocean_proximity
{'INLAND', '<1H OCEAN', 'ISLAND', 'NEAR BAY', 'NEAR OCEAN'}




А) Бинарные (дамми) переменные

In [388]:
df = pd.concat([df, pd.get_dummies(df['ocean_proximity'])], axis=1)
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  \
0       322.0       126.0         8.3252            452600.0        NEAR BAY   
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY   
2       496.0       177.0         7.2574            352100.0        NEAR BAY   
3       558.0       219.0         5.6431            341300.0        NEAR BAY   
4       565.0       259.0         3.8462            342200.0        NEAR BAY   

   bedroom_share  population_per_room  <1H OCEAN  INLAND  ISLAND  NEAR BAY  \
0      14.659091             0.365909      False   False   False      True   
1      15.579659             0.338217      False   False   False      True   
2      12.951602             0.338105      False   False   False      True   
3      18.445840             0.437991      False   False   False      True   
4      17.209588             0.347265      False   False   False      True   

   NEAR OCEAN  
0       False  
1       False  
2       False  
3       False  
4       False

Б) Feature encoding / Target encoding

In [389]:
df_cat = df.groupby('ocean_proximity')['total_rooms'].median()
df_cat = pd.DataFrame(df_cat)
df_cat

total_rooms
ocean_proximity             
<1H OCEAN             2108.0
INLAND                2130.0
ISLAND                1675.0
NEAR BAY              2083.0
NEAR OCEAN            2197.0

In [390]:
df_cat.reset_index(inplace=True)
df_cat

ocean_proximity  total_rooms
0       <1H OCEAN       2108.0
1          INLAND       2130.0
2          ISLAND       1675.0
3        NEAR BAY       2083.0
4      NEAR OCEAN       2197.0

In [391]:
df_cat.rename(columns={'total_rooms': 'median_rooms'}, inplace=True)
df_cat.sort_values(by='median_rooms')

ocean_proximity  median_rooms
2          ISLAND        1675.0
3        NEAR BAY        2083.0
0       <1H OCEAN        2108.0
1          INLAND        2130.0
4      NEAR OCEAN        2197.0

In [392]:
df['total_rooms'].median()

np.float64(2127.0)

In [393]:
df = df.merge(df_cat, on='ocean_proximity')
df.head(3)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   

   population  households  median_income  median_house_value ocean_proximity  \
0       322.0       126.0         8.3252            452600.0        NEAR BAY   
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY   
2       496.0       177.0         7.2574            352100.0        NEAR BAY   

   bedroom_share  population_per_room  <1H OCEAN  INLAND  ISLAND  NEAR BAY  \
0      14.659091             0.365909      False   False   False      True   
1      15.579659             0.338217      False   False   False      True   
2      12.951602             0.338105      False   False   False      True   

   NEAR OCEAN  median_rooms  
0       False        2083.0  
1       False        2083.0  
2       False        2083.0

#### 5.3 А что дальше?

### **latitude, longitude**

In [394]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import geopandas
import plotly.express as px


In [395]:
dn = df.head(100)
dn

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0     -122.23     37.88                41.0        880.0           129.0   
1     -122.22     37.86                21.0       7099.0          1106.0   
2     -122.24     37.85                52.0       1467.0           190.0   
3     -122.25     37.85                52.0       1274.0           235.0   
4     -122.25     37.85                52.0       1627.0           280.0   
..        ...       ...                 ...          ...             ...   
95    -122.26     37.80                36.0       5329.0          2477.0   
96    -122.26     37.82                31.0       4596.0          1331.0   
97    -122.26     37.81                29.0        335.0           107.0   
98    -122.26     37.82                22.0       3682.0          1270.0   
99    -122.26     37.82                37.0       3633.0          1085.0   

    population  households  median_income  median_house_value ocean_proximity  \
0        322.0       126.0         8.3252            452600.0        NEAR BAY   
1       2401.0      1138.0         8.3014            358500.0        NEAR BAY   
2        496.0       177.0         7.2574            352100.0        NEAR BAY   
3        558.0       219.0         5.6431            341300.0        NEAR BAY   
4        565.0       259.0         3.8462            342200.0        NEAR BAY   
..         ...         ...            ...                 ...             ...   
95      3469.0      2323.0         2.0096            130000.0        NEAR BAY   
96      2048.0      1180.0         2.8345            183800.0        NEAR BAY   
97       202.0        91.0         2.0062            125000.0        NEAR BAY   
98      2024.0      1250.0         1.2185            170000.0        NEAR BAY   
99      1838.0       980.0         2.6104            193100.0        NEAR BAY   

    bedroom_share  population_per_room  <1H OCEAN  INLAND  ISLAND  NEAR BAY  \
0       14.659091             0.365909      False   False   False      True   
1       15.579659             0.338217      False   False   False      True   
2       12.951602             0.338105      False   False   False      True   
3       18.445840             0.437991      False   False   False      True   
4       17.209588             0.347265      False   False   False      True   
..            ...                  ...        ...     ...     ...       ...   
95      46.481516             0.650966      False   False   False      True   
96      28.959965             0.445605      False   False   False      True   
97      31.940299             0.602985      False   False   False      True   
98      34.492124             0.549701      False   False   False      True   
99      29.865125             0.505918      False   False   False      True   

    NEAR OCEAN  median_rooms  
0        False        2083.0  
1        False        2083.0  
2        False        2083.0  
3        False        2083.0  
4        False        2083.0  
..         ...           ...  
95       False        2083.0  
96       False        2083.0  
97       False        2083.0  
98       False        2083.0  
99       False        2083.0  

[100 rows x 18 columns]

In [396]:
fig = px.scatter_geo(data_frame=dn, scope='north america', lat='latitude', lon='longitude', size='median_house_value', color='median_house_value', projection='hammer')
fig.update_layout(title_text = 'Дома калифорнии')
fig.show()

ValueError: Mime type rendering requires nbformat>=4.2.0 but it is not installed